lib/api/CDataSummarizationJsonSerializer.cc

```diff
@@ -185,206 +218,181 @@ void CDataSummarizationJsonSerializer::addToJsonStream(TGenericLineWriter& write
     writer.EndObject();
 }
 
-CRetrainableModelJsonDeserializer::TEncoderUPtr
-CRetrainableModelJsonDeserializer::dataSummarizationFromDocumentCompressed(TIStreamSPtr istream,
-                                                                           core::CDataFrame& frame) {
-    rapidjson::IStreamWrapper isw{*istream};
-    rapidjson::Document doc;
-    doc.ParseStream(isw);
-    if (doc.HasMember(JSON_COMPRESSED_DATA_SUMMARIZATION_TAG) == false ||
-        doc[JSON_COMPRESSED_DATA_SUMMARIZATION_TAG].IsObject() == false) {
-        LOG_ERROR(<< "Field " << JSON_COMPRESSED_DATA_SUMMARIZATION_TAG
-                  << " not found or is not an object.");
+CRetrainableModelJsonReader::TEncoderUPtr
+CRetrainableModelJsonReader::dataSummarizationFromJsonStream(TIStreamSPtr istream,
+                                                             core::CDataFrame& frame) {
+    if (istream == nullptr) {
         return nullptr;
     }
-    auto& compressedDataSummarization = doc[JSON_COMPRESSED_DATA_SUMMARIZATION_TAG];
-    if (compressedDataSummarization.HasMember(JSON_DATA_SUMMARIZATION_TAG) == false ||
-        compressedDataSummarization[JSON_DATA_SUMMARIZATION_TAG].IsString() == false) {
-        LOG_ERROR(<< "Field " << JSON_DATA_SUMMARIZATION_TAG << " not found or is not a string.");
-        return nullptr;
+    try {
+        return dataSummarizationFromJson(*istream, frame);
+    } catch (...) {
```

Comment:
This is silently eating `std::bad_alloc`.  It's also silently eating Windows structured exceptions that Windows APIs might throw (only `...` catches these using the C++ syntax).

We are in a strange situation because our coding guidelines say exceptions shouldn't be used, so we assume that in the places where we _do_ use them that we can be very simplistic in how we do it.  But the more places where we use them the less appropriate the simplistic handling becomes.

If you only expect this to ever catch `std::runtime_exception` then catching that explicitly would solve the problems of eating `std::bad_alloc` and Windows structured exceptions.  If you're expecting other types of exception like `std::logic_error` or `std::bad_cast` then it's more painful to handle nicely.

Reply:
Ok that seems reasonable. However, I'd add that in this case any exception being thrown should probably trigger returning a null summarisation which causes the program to call terminate elsewhere. This is probably slightly safer IMO than just letting the exception propagate since, as you say, we don't in general know if it might be caught elsewhere (although in this case I'm pretty sure it isn't).

I actually think for restore throwing exceptions is a good way of handling bad or missing state (it made this error checking code much cleaner to write) and we want to fail as fast as possible. So my preferred solution would be to allow any restore code to throw (including all error handling where we currently return false from restore methods) have no try catch in any individual class restore and wrap all top level calls to read state with try catch. That is beyond the scope of this change, but I guess having targeted catch here is more in keeping with that approach.

Reply:
Done in [f2c3588](https://github.com/elastic/ml-cpp/pull/1894/commits/f2c35881e9ec52f95d64cd4ba16bc45c3712fed3).